### PR TITLE
[overlay] Support the older @rpath Darwin library for _Builtin_float's $ld$previous$ symbols

### PR DESCRIPTION
**Pull request by ian-twilightcoder**
The _Builtin_float symbols were moved twice, most recently from the OS Darwin library, but previously they were in the embedded @rpath Darwin library. @_originallyDefinedIn doesn't support multiple install names, but it can be replaced with -module-abi-name, and then multiple $ld$previous$ symbols can be used. Update the Platform and Concurrency magic symbols to use $ld$previous$ everywhere.

rdar://130107191

### CODE REVIEW

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test

**Comment by ian-twilightcoder** [comment]
@swift-ci please smoke test

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test

File: stdlib/public/Concurrency/linker-support/magic-symbols-for-install-name.c [hunk 1]
```diff
@@ -22,116 +22,55 @@
...
-#include <Availability.h>
+#include <mach-o/loader.h>
 #include <TargetConditionals.h>
 #include "swift/shims/Visibility.h"
 
-#define RPATH_INSTALL_NAME_DIRECTIVE_IMPL2(name, major, minor) \
-  SWIFT_RUNTIME_EXPORT const char install_name_ ## major ## _ ## minor \
-  __asm("$ld$install_name$os" #major "." #minor "$@rpath/lib" #name ".dylib"); \
-  const char install_name_ ## major ## _ ## minor = 0;
+// Concurrency was supported as an embedded library in macOS (née OS X) 10.9, iOS 7.0, watchOS 2.0, tvOS 9.0.
```
Comment (DougGregor) [inline stdlib/public/Concurrency/linker-support/magic-symbols-for-install-name.c:29]: Technically, Concurrency only back-deploys as an embedded library to macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0

File: test/abi/macOS/arm64/builtin_float.swift [hunk 2]
```diff
@@ -0,0 +1,37 @@
+// RUN: %empty-directory(%t)
+// RUN: %llvm-nm -g --defined-only -f just-symbols %stdlib_dir/arm64/libswift_Builtin_float.dylib > %t/symbols
+// RUN: %abi-symbol-checker %s %t/symbols
+// RUN: diff -u %S/../../Inputs/macOS/arm64/builtin_float/baseline %t/symbols
```
Comment (DougGregor) [inline test/abi/macOS/arm64/builtin_float.swift:4]: Thank you for adding these baseline tests!

File: stdlib/linker-support/magic-symbols-for-install-name.c [hunk 3]
```diff
@@ -19,74 +19,54 @@
...
 
-#define RPATH_INSTALL_NAME_DIRECTIVE_IMPL(name, major, minor) \
-  RPATH_INSTALL_NAME_DIRECTIVE_IMPL2(name, major, minor)
+// The linker uses a specially formatted symbol to do the back deployment:
+// $ld$previous$<install-name>$<compatibility-version>$<platform>$<start-version>$<end-version>$<symbol-name>$
+// compatibility-version and symbol-name are left off to apply to all library versions and symbols.
+// This symbol isn't a legal C identifier, so it needs to be specified with __asm.
+#define RPATH_PREVIOUS_DIRECTIVE_IMPL(name, platform, startVersion, endVersion) \
+  SWIFT_RUNTIME_EXPORT const char ld_previous_ ## platform \
+  __asm("$ld$previous$@rpath/lib" __STRING(name) ".dylib$$" __STRING(platform) "$" __STRING(startVersion) "$" __STRING(endVersion) "$$"); \
```
Comment (DougGregor) [inline stdlib/linker-support/magic-symbols-for-install-name.c:40]: I love the switch to `$ld$previous$`, it makes everything so much cleaner.

File: stdlib/public/Concurrency/linker-support/magic-symbols-for-install-name.c [hunk 4]
```diff
@@ -22,116 +22,55 @@
...
-  RPATH_INSTALL_NAME_DIRECTIVE(11,  5)
-  RPATH_INSTALL_NAME_DIRECTIVE(11,  6)
-
- // Link against @rpath/libswift_Concurrency.dylib for macCatalyst < 15.0.
-SWIFT_RUNTIME_EXPORT const char ld_previous_macCatalyst
-__asm("$ld$previous$@rpath/libswift_Concurrency.dylib$$6$1.0$15.0$");
-
-const char ld_previous_macCatalyst = 0;
+#if TARGET_OS_OSX || TARGET_OS_MACCATALYST
+RPATH_PREVIOUS_DIRECTIVE(PLATFORM_MACOS, 10.9, 12.0)
```
Comment (DougGregor) [inline stdlib/public/Concurrency/linker-support/magic-symbols-for-install-name.c:52]: I think these versions should reflect the the back-deployment versions for concurrency: macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0

**Review by DougGregor — COMMENTED** [review]
The concurrency back-deployment versions aren't quite right (see comments), but otherwise this LGTM

File: stdlib/public/Concurrency/linker-support/magic-symbols-for-install-name.c (hunk 1 above)
Comment (ian-twilightcoder) [reply stdlib/public/Concurrency/linker-support/magic-symbols-for-install-name.c:29]: Ah, I figured the original code was probably wrong to back deploy all the way to the original Swift versions.

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test

**Review by DougGregor — APPROVED** [review]
LGTM, thanks!

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test

**Review by mikeash — APPROVED** [review]

**Pull request merged**